service/session.hh

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "utils/UUID.hh"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/shared_future.hh>
+
+#include <boost/intrusive/list.hpp>
+
+namespace service {
+
+using session_id = utils::UUID;
+
+/// Session is used to track execution of work related to some greater task, identified by session_id.
```

Comment:
I think this comment needs to explain why session exists in addition to fencing, and how they operate on a different level. I also wonder whether it is a scylla concept or a lower level concept like seastar's. Essentially the way it is implemented here a sesison is a kind of an identifiable/findable gate. The whole persistence story exists outside session definition - which is, interestingly, suggests that sessions don't have to be persistent.

Reply:
For example, I would mention here that sessions are used to track long running RPCs based interactions which potentially involve *multiple* nodes and use many one-way RPCs to accomplish a task. That we purposefully want session identifiers to be independent from e.g. topology version, in order to keep some sessions open when a topology changes, and, vice versa, close some sessions without changing the cluster topology.

Reply:
> which is, interestingly, suggests that sessions don't have to be persistent.

And frankly I do not see why this series bother to persist them. A session is always created in a state S1 and always removed in a state S2. S2 will never move to S3 unless the session is closed which means that session identifiers do not have to be persisted and can be chosen by each node randomly while preserving in memory mapping tablet_id->session_id. The code that looks for a session_id in tablet metadata will lookup it instead by calling `session_id get_streaming_session_id_for_tablet(tablet_id)`

Reply:
> I think this comment needs to explain why session exists in addition to fencing, and how they operate on a different level. I also wonder whether it is a scylla concept or a lower level concept like seastar's. Essentially the way it is implemented here a sesison is a kind of an identifiable/findable gate.

This is a low level concept, so I think the description of the big picture doesn't belong in a comment here. Perhaps a docs/dev/topology-guard.md, and link to it.

Reply:
> And frankly I do not see why this series bother to persist them. A session is always created in a state S1 and always removed in a state S2. S2 will never move to S3 unless the session is closed which means that session identifiers do not have to be persisted and can be chosen by each node randomly while preserving in memory mapping tablet_id->session_id. The code that looks for a session_id in tablet metadata will lookup it instead by calling `session_id get_streaming_session_id_for_tablet(tablet_id)`

Each node has to use the same session id so that you can pass the guard across nodes. If each node uses different session id, how do you detect if the guard is still valid?

Reply:
Nodes do not just start streaming no? There is a node that contacts all other nodes and asks to stream data to it. This node chooses a random id for the session and pass it to all pears together with the streaming request. Isn't this what the patch "Propagate topology_guard to receivers " essentially doing?